README.md
## elm-upgrade [![](https://img.shields.io/npm/v/elm-upgrade.svg)](https://www.npmjs.com/package/elm-upgrade) [![Build Status](https://travis-ci.org/avh4/elm-upgrade.svg?branch=master)](https://travis-ci.org/avh4/elm-upgrade)

`elm-upgrade` can help you **upgrade your Elm 0.18 projects to Elm 0.19**.  It attempts to automate many of the steps in the [Elm 0.19 upgrade guide][upgrade].  `elm-upgrade` will do the following:
  - Convert your `elm-package.json` file to ...
    - ... an application `elm.json` if your project has no exposed modules
    - ... a package `elm.json` if your project has at least one exposed module
  - Try to upgrade all of your project dependencies
  - Warn you if some of your project dependencies don't support Elm 0.19 yet
  - Use [elm-format](https://github.com/avh4/elm-format) `--upgrade` to upgrade your code, which includes the following:
    - Convert escaped characters in strings to the new syntax (`\u{xxxx}`)
    - Inline uses of functions which were removed in Elm 0.19:
      - `(,,)`, `(,,,)`, etc tuple constructor functions
      - `Platform.Cmd.(!)`
      - `flip`, `curry`, `uncurry`, and `rem` from the `Basics` module
    - Upgrade code that uses `Html.Attributes.style`

`elm-upgrade` can also **upgrade dependencies of your Elm 0.19 applications**.
If you are already using Elm 0.19, `elm-upgrade` will to the following:
  - check for newer versions of all your direct dependencies and try to install them


## How to use **elm-upgrade**

First install Elm 0.19.1 and the latest version of [elm-format](https://github.com/avh4/elm-format#installation-).  (If you want to install them locally for your project, you can do so with the following: )

```sh
cd path/to/my/elm/project
npm install elm@latest-0.19.1
npm install elm-format@latest-0.19.1
```

Then run the following in your terminal:

```sh
cd path/to/my/elm/project
npx elm-upgrade@latest
```

NOTE:  `npx` ships with node 8.2 and later.  If you need to use an older version of node, you can still use elm-upgrade with `npm install -g elm-upgrade; elm-upgrade`.

After the automated upgrade, you will probably still have to fix a few things.  See the [Elm 0.19 upgrade guide][upgrade] for more details.

[upgrade]: https://github.com/elm/compiler/blob/master/docs/upgrade-instructions/0.19.0.md

## What it looks like

```
$ elm-upgrade
INFO: Found elm at /Users/avh4/workspace/elm-upgrade/node_modules/.bin/elm
INFO: Found elm 0.19.0
INFO: Found elm-format at /Users/avh4/workspace/elm-upgrade/node_modules/.bin/elm-format
INFO: Found elm-format 0.8.0
INFO: Cleaning ./elm-stuff before upgrading
INFO: Converting elm-package.json -> elm.json
INFO: Detected an application project (this project has no exposed modules)
INFO: Switching from elm-lang/core (deprecated) to elm/core
INFO: Installing latest version of elm/core
It is already installed!
INFO: Detected use of elm-lang/core#Random; installing elm/random
Here is my plan:

  Add:
    elm/random    1.0.0
    elm/time      1.0.0

Would you like me to update your elm.json accordingly? [Y/n]: y
Dependencies loaded from local cache.
Dependencies ready!
INFO: Detected use of elm-lang/core#Time; installing elm/time
I found it in your elm.json file, but in the "indirect" dependencies.
Should I move it into "direct" dependencies for more general use? [Y/n]: y
Dependencies loaded from local cache.
Dependencies ready!
INFO: Switching from elm-lang/html (deprecated) to elm/html
INFO: Installing latest version of elm/html
Here is my plan:

  Add:
    elm/html           1.0.0
    elm/virtual-dom    1.0.0

Would you like me to update your elm.json accordingly? [Y/n]: y
Dependencies loaded from local cache.
Dependencies ready!
INFO: Upgrading *.elm files in ./


SUCCESS! Your project's dependencies and code have been upgraded.
However, your project may not yet compile due to API changes in your
dependencies.

See <https://github.com/elm/compiler/blob/master/upgrade-docs/0.19.md>
and the documentation for your dependencies for more information.

Here are some common upgrade steps that you will need to do manually:

- elm/core
  - [ ] Replace uses of toString with String.fromInt, String.fromFloat, or Debug.toString as appropriate
- elm/html
  - [ ] If you used Html.program*, install elm/browser and switch to Browser.element or Browser.document
  - [ ] If you used Html.beginnerProgram, install elm/browser and switch Browser.sandbox

$ git add -N elm.json
$ git diff
```
```diff
diff --git a/Main.elm b/Main.elm
index 7dd0dfb..1cbcdea 100644
--- a/Main.elm
+++ b/Main.elm
@@ -1,4 +1,4 @@
-module Main exposing (..)
+module Main exposing (Model, Msg(..), height, init, main, update, view, width)

 import Html exposing (..)
 import Html.Attributes exposing (..)
@@ -122,31 +122,27 @@ view model =
         cursorY =
             toString model.yPosition ++ "px"
     in
-    div [ style [ ( "position", "relative" ) ] ]
+    div [ style "position" "relative" ]
         [ img
-            [ style
-                [ ( "width", "100%" )
-                , ( "max-width", toString width ++ "px" )
-                , ( "margin-left", "-50%" )
-                , ( "position", "absolute" )
-                , ( "left", "50%" )
-                ]
+            [ style "width" "100%"
+            , style "max-width" (toString width ++ "px")
+            , style "margin-left" "-50%"
+            , style "position" "absolute"
+            , style "left" "50%"
             , src "assets/oujia_6.jpeg"
             ]
             []
         ]
diff --git a/elm-package.json b/elm-package.json
deleted file mode 100644
index f5ba1c5..0000000
--- a/elm-package.json
+++ /dev/null
@@ -1,15 +0,0 @@
-{
-    "version": "1.0.0",
-    "summary": "helpful summary of your project, less than 80 characters",
-    "repository": "https://github.com/user/project.git",
-    "license": "BSD3",
-    "source-directories": [
-        "."
-    ],
-    "exposed-modules": [],
-    "dependencies": {
-        "elm-lang/core": "5.1.1 <= v < 6.0.0",
-        "elm-lang/html": "2.0.0 <= v < 3.0.0"
-    },
-    "elm-version": "0.18.0 <= v < 0.19.0"
-}
diff --git a/elm.json b/elm.json
index e69de29..65d31f3 100644
--- a/elm.json
+++ b/elm.json
@@ -0,0 +1,23 @@
+{
+    "type": "application",
+    "source-directories": [
+        "."
+    ],
+    "elm-version": "0.19.0",
+    "dependencies": {
+        "direct": {
+            "elm/core": "1.0.0",
+            "elm/html": "1.0.0",
+            "elm/random": "1.0.0",
+            "elm/time": "1.0.0"
+        },
+        "indirect": {
+            "elm/json": "1.0.0",
+            "elm/virtual-dom": "1.0.0"
+        }
+    },
+    "test-dependencies": {
+        "direct": {},
+        "indirect": {}
+    }
+}
\ No newline at end of file
```


## Development info for contributors to elm-upgrade

See [CONTRIBUTING.md](CONTRIBUTING.md).


CHANGELOG.md
# 0.19.7

- Upgrading applications now works when you have Elm 0.19.1 (thanks to @ChristophP)


# (changelog not updated through 0.19.6)


# 0.18.7

 - Support for systems that use an HTTP proxy
 - New automatic dependency renames:
   - NoRedInk/elm-string-extra -> elm-community/string-extra
   - lorenzo/elm-string-addons -> elm-community/string-extra


tests/make_links.sh
#!/bin/bash

function make_link {
  if which -s "$1"; then
    ln -sfv "$(which "$1")" "$BASE/bin/"
  else
    echo "$0: ERROR: not found on \$PATH: $1"
    exit 1
  fi
}

echo "Locating command line tools needed for integration tests..."
BASE="${BASH_SOURCE%/*}"
make_link node
make_link rsync
make_link git
ln -sfv "$(which security)" "$BASE/bin/" || true # required by elm, but only on Mac OS
make_link jq
make_link mv
make_link yes

echo


src/packageTransformations.js
module.exports = {
  "NoRedInk/elm-decode-pipeline": [
    {
      action: "installPackage",
      packageName: "NoRedInk/elm-json-decode-pipeline",
      todos: [
        "Changes uses of Json.Decode.Pipeline.decode to Json.Decode.succeed"
      ]
    }
  ],
  "elm-community/elm-test": [
    { action: "installPackage", packageName: "elm-explorations/test" }
  ],
  "elm-lang/animation-frame": [
    {
      action: "installPackage",
      packageName: "elm/browser",
      todos: [
        "Change code using AnimationFrame.* to use Browser.Events.onAnimationFrame*"
      ]
    }
  ],
  "elm-lang/core": [
    {
      action: "installPackage",
      packageName: "elm/core",
      todos: [
        "Replace uses of toString with String.fromInt, String.fromFloat, or Debug.toString as appropriate"
      ]
    },
    {
      action: "match",
      condition: {
        type: "usesModule",
        modules: ["Json.Decode", "Json.Encode"]
      },
      ifMet: [{ action: "installPackage", packageName: "elm/json" }]
    },
    {
      action: "match",
      condition: {
        type: "usesModule",
        modules: ["Random"]
      },
      ifMet: [{ action: "installPackage", packageName: "elm/random" }]
    },
    {
      action: "match",
      condition: {
        type: "usesModule",
        modules: ["Time", "Date"]
      },
      ifMet: [
        {
          action: "installPackage",
          packageName: "elm/time",
          todos: [
            "Read the new documentation here: https://package.elm-lang.org/packages/elm/time/latest/",
            "Replace uses of Date and Time with Time.Posix"
          ]
        }
      ]
    },
    {
      action: "match",
      condition: {
        type: "usesModule",
        modules: ["Regex"]
      },
      ifMet: [{ action: "installPackage", packageName: "elm/regex" }]
    }
  ],
  "elm-lang/html": [
    {
      action: "installPackage",
      packageName: "elm/html",
      todos: [
        "If you used Html.program*, install elm/browser and switch to Browser.element or Browser.document",
        "If you used Html.beginnerProgram, install elm/browser and switch Browser.sandbox"
      ]
    }
  ],
  "elm-lang/http": [{ action: "installPackage", packageName: "elm/http" }],
  "elm-lang/navigation": [
    {
      action: "installPackage",
      packageName: "elm/browser",
      todos: [
        "Change code using Navigation.program* to use Browser.application",
        "Use the Browser.Key passed to your init function in any calls to Browser.Navigation.pushUrl/replaceUrl/back/forward"
      ]
    },
    {
      action: "installPackage",
      packageName: "elm/url",
      todos: ["Changes uses of Navigation.Location to Url.Url"]
    }
  ],
  "elm-lang/svg": [{ action: "installPackage", packageName: "elm/svg" }],
  "elm-lang/window": [
    {
      action: "installPackage",
      packageName: "elm/browser",
      todos: ["Change code using Window.* to use Browser.Events.onResize"]
    }
  ],
  "elm-lang/virtual-dom": [
    { action: "installPackage", packageName: "elm/virtual-dom" }
  ],
  "elm-tools/parser": [{ action: "installPackage", packageName: "elm/parser" }],
  "evancz/elm-markdown": [
    { action: "installPackage", packageName: "elm-explorations/markdown" }
  ],
  "evancz/url-parser": [
    {
      action: "installPackage",
      packageName: "elm/url",
      todos: ["Change code using UrlParser.* to use Url.Parser.*"]
    }
  ],
  "mgold/elm-random-pcg": [
    {
      action: "installPackage",
      packageName: "elm/random",
      todos: ["Change references to Random.Pcg.* to Random.*"]
    }
  ],
  "ohanhi/keyboard-extra": [
    { action: "installPackage", packageName: "ohanhi/keyboard" }
  ],
  "thebritican/elm-autocomplete": [
    { action: "installPackage", packageName: "ContaSystemer/elm-menu" }
  ],
  "elm-community/linear-algebra": [
    { action: "installPackage", packageName: "elm-explorations/linear-algebra" }
  ],
  "elm-community/webgl": [
    { action: "installPackage", packageName: "elm-explorations/webgl" }
  ],
  "elm-lang/keyboard": [
    { action: "installPackage", packageName: "elm/browser" }
  ],
  "elm-lang/dom": [{ action: "installPackage", packageName: "elm/browser" }],
  "mpizenberg/elm-mouse-events": [
    {
      action: "installPackage",
      packageName: "mpizenberg/elm-pointer-events",
      todos: [
        "Read the upgrade instructions here: https://github.com/mpizenberg/elm-pointer-events/blob/master/upgrade.md"
      ]
    }
  ],
  "mpizenberg/elm-touch-events": [
    {
      action: "installPackage",
      packageName: "mpizenberg/elm-pointer-events",
      todos: [
        "Read the upgrade instructions here: https://github.com/mpizenberg/elm-pointer-events/blob/master/upgrade.md"
      ]
    }
  ],
  "ryannhg/elm-date-format": [
    { action: "installPackage", packageName: "ryannhg/date-format" }
  ],
  "rtfeldman/hex": [
    { action: "installPackage", packageName: "rtfeldman/elm-hex" }
  ],
  "elm-lang/mouse": [{ action: "installPackage", packageName: "elm/browser" }],
  "avh4/elm-transducers": [
    {
      action: "installPackage",
      packageName: "avh4-experimental/elm-transducers"
    }
  ],
  "dillonkearns/graphqelm": [
    {
      action: "installPackage",
      packageName: "dillonkearns/elm-graphql",
      todos: ["Changes references to Graphqelm.* to Graphql.*"]
    }
  ],
  "BrianHicks/elm-benchmark": [
    { action: "installPackage", packageName: "elm-explorations/benchmark" }
  ]
};


upgrade.js
#!/usr/bin/env node

var fs = require("fs-extra");
var childProcess = require("child_process");
var which = require("which");
var semver = require("semver");
var path = require("path");
var safename = require("safename");

var packageHost = "https://package.elm-lang.org";

var packageTransformations = require("./src/packageTransformations");

var logHandle = null;

function exit(status) {
  if (logHandle) {
    fs.closeSync(logHandle);
  }
  process.exit(status);
}

function logFile(message) {
  if (logHandle == null) {
    var timestamp = new Date().toISOString();
    var filename = safename("elm-upgrade-" + timestamp + ".log");
    logHandle = fs.openSync(filename, "ax");
  }
  fs.writeSync(logHandle, message);
}

function logMessage(message) {
  process.stdout.write(message);
  logFile(message);
}

function logErrorMessage(message) {
  process.stderr.write(message);
  logFile(message);
}

function logInfo(message) {
  logMessage("INFO: " + message + "\n");
}

function logWarning(message) {
  logMessage("WARNING: " + message + "\n");
}

function logError(message) {
  logErrorMessage("ERROR: " + message + "\n");
}

function howToInstallElm() {
  return "Install Elm here https://guide.elm-lang.org/get_started.html#install\n";
}

function howToInstallElmFormat() {
  return "You can download Elm format here https://github.com/avh4/elm-format#installation-\n";
}

function displayHintForNonUpgradedPackage(packageName) {
  logWarning(`${packageName} has not been upgraded to 0.19 yet!`);
}

function displaySuccessMessage(packagesRequiringUpgrade) {
  logMessage(
    "SUCCESS! Your project's dependencies and code have been upgraded.\n" +
      "However, your project may not yet compile due to API changes in your\n" +
      "dependencies.\n\n" +
      "See <https://github.com/elm/compiler/blob/master/upgrade-docs/0.19.md>\n" +
      "and the documentation for your dependencies for more information.\n\n"
  );

  if (packagesRequiringUpgrade.length > 0) {
    logMessage(
      // "WARNING! " +
      //   packagesRequiringUpgrade.length +
      //   " of your dependencies have not yet been upgraded to\n" +
      //   "support Elm 0.19. You can create an issue to request the packages be\n" +
      //   "upgraded here:\n"
      "WARNING! " +
        packagesRequiringUpgrade.length +
        " of your dependencies have not yet been upgraded to\n" +
        "support Elm 0.19.\n"
    );
    packagesRequiringUpgrade.forEach(function(packageName) {
      logMessage(
        // "  - https://github.com/" + packageName + "/issues\n"
        "  - https://github.com/" + packageName + "\n"
      );
    });
    logMessage("\n");
  }
}
/*
  Locate a binary based on name, if not found, error out with message
Provide binFolder in order to look there first
*/
function findBinary(binFolder, name, message, extraInfo) {
  var binary = null;

  // first look in binFolder
  try {
    binary = which.sync(name, { path: binFolder });
  } catch (e) {}

  // then just look all over path
  if (binary === null) {
    try {
      binary = which.sync(name);
    } catch (e) {
      logError(message);
      logErrorMessage(extraInfo);
      exit(1);
    }
  }

  logInfo("Found " + name + " at " + binary);
  return binary;
}

function modifyElmJsonSync(transform) {
  var elmJson = JSON.parse(fs.readFileSync("elm.json", "utf8"));
  elmJson = transform(elmJson);
  saveElmJson(elmJson);
}

function saveElmJson(elmPackage) {
  fs.writeFileSync("elm.json", JSON.stringify(elmPackage, null, 4));
}

function findInFiles(roots, patterns) {
  return roots.some(function(root) {
    return fs.readdirSync(root).some(function(file) {
      var filename = path.join(root, file);
      if (file.slice(-4) === ".elm") {
        var contents = fs.readFileSync(filename, "utf8");
        return patterns.some(function(pattern) {
          return contents.match(pattern);
        });
      } else if (fs.statSync(filename).isDirectory()) {
        return findInFiles([filename], patterns);
      } else {
        return false;
      }
    });
  });
}

function main(knownPackages) {
  var supportedPackages = {};
  knownPackages.forEach(function(packageInfo) {
    supportedPackages[packageInfo.name] = packageInfo.versions.slice(-1)[0];
  });

  function supportsElm0_19(packageName) {
    return !!supportedPackages[packageName];
  }

  function latestVersion(packageName) {
    return supportedPackages[packageName];
  }

  var localBinPath = "./node_modules/.bin/";
  var localFindBinary = findBinary.bind(null, localBinPath);

  var elm = localFindBinary(
    "elm",
    "elm was not found on your PATH.  Make sure you have Elm 0.19 installed.",
    howToInstallElm()
  );

  var elmUsage = childProcess.execFileSync(elm, ["--version"]);
  var elmVersion = elmUsage.toString().split("\n")[0];
  if (!elmVersion.match(/^0\.19\./)) {
    logErrorMessage(
      "ERROR: Elm 0.19 required, but found " +
        elmVersion +
        "\n" +
        howToInstallElm()
    );
    exit(1);
  }
  logInfo("Found elm " + elmVersion);

  var elmFormat = localFindBinary(
    "elm-format",
    "elm-format was not found on your PATH.  Make sure you have elm-format installed.",
    howToInstallElmFormat()
  );

  var elmFormatUsage = childProcess.execFileSync(elmFormat);
  var elmFormatVersion = elmFormatUsage
    .toString()
    .split("\n")[0]
    .trim()
    .split(" ")[1];
  if (semver.lt(elmFormatVersion, "0.8.0")) {
    logError("elm-format >= 0.8.0 required, but found " + elmFormatVersion);
    logErrorMessage(howToInstallElmFormat());
    exit(1);
  }
  logInfo("Found elm-format " + elmFormatVersion);

  function installPackage(name) {
    try {
      childProcess.execFileSync(elm, ["install", name], {
        stdio: "inherit"
      });
    } catch (e) {
      logWarning(`Failed to upgrade ${name}!`);
    }
  }

  if (fs.existsSync("elm.json")) {
    var elmJson = JSON.parse(fs.readFileSync("elm.json", "utf8"));
    var isPackage = elmJson.type == "package";

    if (isPackage) {
      logMessage(
        "\n" +
          "***\n" +
          "*** ./elm.json already exists.\n" +
          "*** It looks like this project has already been upgraded to Elm 0.19.\n" +
          "*** Since this is a package project, you should keep the version bounds\n" +
          "*** for your dependencies as wide as possible.\n" +
          "***\n" +
          "\n\n"
      );
      logInfo("Checking if all your dependencies support Elm 0.19...");

      var foundBadPackage = false;
      var packages = Object.keys(elmJson.dependencies);
      packages.forEach(function(packageName) {
        var latestVersion_ = latestVersion(packageName);
        if (!latestVersion_) {
          foundBadPackage = true;
          displayHintForNonUpgradedPackage(packageName);
        }
      });

      if (!foundBadPackage) {
        logMessage(
          "\n\n" + "SUCCESS! Your project's dependencies look good.\n\n"
        );
      }

      exit(0);
    } else {
      logMessage(
        "\n" +
          "***\n" +
          "*** ./elm.json already exists.\n" +
          "*** It looks like this project has already been upgraded to Elm 0.19.\n" +
          "*** Would you like me to upgrade your project's dependencies?\n" +
          "***\n" +
          "\n"
      );
      var yn = require("yn");
      var readline = require("readline");

      var rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
      });

      var updateElmVersion = function() {
        modifyElmJsonSync(function(elmJson) {
          if (elmJson["elm-version"] != elmVersion) {
            logInfo("Updating elm-version to " + elmVersion);
          }
          elmJson["elm-version"] = elmVersion;
          return elmJson;
        });
      };
      var updatePackages = function() {
        var packages = Object.keys(elmJson.dependencies.direct);
        packages.forEach(function(packageName) {
          var currentVersion = elmJson.dependencies.direct[packageName];
          var latestVersion_ = latestVersion(packageName);
          if (!latestVersion_) {
            displayHintForNonUpgradedPackage(packageName);
          } else if (semver.lt(currentVersion, latestVersion_)) {
            logInfo("Installing latest version of " + packageName);
            modifyElmJsonSync(function(elmJson) {
              delete elmJson.dependencies.direct[packageName];
              return elmJson;
            });
            installPackage(packageName);
          }
        });
        logMessage(
          "\n\n" +
            "SUCCESS! Your project's dependencies have been upgraded.\n" +
            "However, your project may not yet compile due to API changes in your\n" +
            "dependencies.\n\n"
        );
      };

      rl.question("[Y/n]: ", answer => {
        logMessage("\n");
        var proceed = yn(answer);
        if (proceed) {
          updateElmVersion();
          updatePackages();
        }
        rl.close();
        exit(0);
      });
    }
    return;
  }

  if (!fs.existsSync("elm-package.json")) {
    logError(
      "You must run the upgrade from a folder containing elm-package.json"
    );
    exit(1);
  }

  // TODO: Warning and prompt if git is not being used
  // TODO: Error if git workspace is dirty

  var elmPackage = JSON.parse(fs.readFileSync("elm-package.json", "utf8"));

  if (!elmPackage["elm-version"].startsWith("0.18.")) {
    logError(
      "This is not an Elm 0.18 project.  Current project uses Elm " +
        elmPackage["elm-version"]
    );
    exit(1);
  }

  logInfo("Cleaning ./elm-stuff before upgrading");
  fs.removeSync("elm-stuff");

  var packageName = elmPackage["repository"].match(
    /^https:\/\/github.com\/([^/]*\/[^/]*)\.git$/
  )[1];
  // TODO: Error if packageName doesn't parse

  logInfo("Converting elm-package.json -> elm.json");

  var isPackage = elmPackage["exposed-modules"].length > 0;

  var elmJson;
  if (isPackage) {
    logInfo("Detected a package project (this project has exposed modules)");

    if (elmPackage["license"] === "BSD3") {
      logInfo(
        "Detected 'BSD3' license, which is not a valid SPDX license identifier; converting to 'BSD-3-Clause'"
      );
      elmPackage["license"] = "BSD-3-Clause";
    }

    elmJson = {
      type: "package",
      name: packageName,
      summary: elmPackage["summary"],
      license: elmPackage["license"],
      version: elmPackage["version"],
      "exposed-modules": elmPackage["exposed-modules"],
      "elm-version": "0.19.0 <= v < 0.20.0",
      dependencies: {
        "elm/core": "1.0.0 <= v < 2.0.0"
      },
      "test-dependencies": {}
    };
  } else {
    logInfo(
      "Detected an application project (this project has no exposed modules)"
    );
    elmJson = {
      type: "application",
      "source-directories": elmPackage["source-directories"],
      "elm-version": elmVersion,
      dependencies: {
        direct: {
          "elm/core": "1.0.2"
        },
        indirect: {
          "elm/json": "1.1.3"
        }
      },
      "test-dependencies": {
        direct: {},
        indirect: {}
      }
    };
  }
  saveElmJson(elmJson);

  var packagesToInstall = Object.keys(elmPackage["dependencies"]);

  var packagesRequiringUpgrade = [];
  var todos = {};

  packagesToInstall.forEach(function(packageName) {
    var oldPackageName = packageName;
    var actions = packageTransformations[packageName] || [
      { action: "keepPackage", packageName: packageName }
    ];
    actions.forEach(function(action) {
      switch (action.action) {
        case "installPackage":
          var newPackageName = action.packageName;
          logInfo(
            "Switching from " +
              packageName +
              " (deprecated) to " +
              newPackageName
          );

          if (!supportsElm0_19(newPackageName)) {
            displayHintForNonUpgradedPackage(newPackageName);
            if (isPackage) {
              // TODO: not tested
              packagesRequiringUpgrade[newPackageName] =
                elmPackage.dependencies[oldPackageName];
            } else {
              packagesRequiringUpgrade[
                newPackageName
              ] = elmPackage.dependencies[oldPackageName].split(" ")[0];
            }
            return;
          }

          todos[newPackageName] = (todos[newPackageName] || []).concat(
            action.todos || []
          );

          logInfo("Installing latest version of " + newPackageName);
          installPackage(newPackageName);
          break;

        case "keepPackage":
          var newPackageName = action.packageName;
          if (!supportsElm0_19(newPackageName)) {
            displayHintForNonUpgradedPackage(newPackageName);
            if (isPackage) {
              // TODO: not tested
              packagesRequiringUpgrade[newPackageName] =
                elmPackage.dependencies[oldPackageName];
            } else {
              packagesRequiringUpgrade[
                newPackageName
              ] = elmPackage.dependencies[oldPackageName].split(" ")[0];
            }
            return;
          }

          logInfo("Installing latest version of " + newPackageName);
          installPackage(newPackageName);
          break;

        case "match":
          var resultingActions = [];
          switch (action.condition.type) {
            case "usesModule":
              var moduleNames = action.condition.modules;
              for (var i = 0; i < moduleNames.length; i++) {
                var moduleName = moduleNames[i];
                if (
                  findInFiles(elmPackage["source-directories"], [
                    RegExp("(^|[\n\r])import " + moduleName + "[ \n\r]")
                  ])
                ) {
                  resultingActions = action.ifMet;
                  break;
                }
              }
              break;

            default:
              throw "ERROR: Unexpected condition: " +
                JSON.stringify(action.condition);
          }

          resultingActions.forEach(function(action) {
            if (action.action !== "installPackage") {
              throw "ERROR: Not yet implemented (in match action): " +
                JSON.stringify(action);
            }
            var target = action.packageName;

            todos[newPackageName] = (todos[newPackageName] || []).concat(
              action.todos || []
            );

            logInfo(
              "Detected use of " +
                oldPackageName +
                "#" +
                moduleName +
                "; installing " +
                target
            );
            installPackage(target);
          });
          break;

        default:
          throw "ERROR: Unexpected action: " + JSON.stringify(action);
      }
    });
  });

  elmJson = JSON.parse(fs.readFileSync("elm.json", "utf8"));
  Object.keys(packagesRequiringUpgrade).forEach(function(name) {
    if (isPackage) {
      elmJson.dependencies[name] = packagesRequiringUpgrade[name];
    } else {
      elmJson.dependencies.direct[name] = packagesRequiringUpgrade[name];
    }
  });
  saveElmJson(elmJson);

  fs.unlinkSync("elm-package.json");

  // TODO: deal with source-directories for packages

  elmPackage["source-directories"].forEach(function(sourceDir) {
    if (!fs.existsSync(sourceDir)) {
      logWarning(
        "source directory " +
          sourceDir +
          " listed in your elm-package.json does not exist"
      );
    } else {
      logInfo("Upgrading *.elm files in " + sourceDir + "/");
      childProcess.execFileSync(elmFormat, [
        "--upgrade",
        "--yes",
        "--elm-version",
        "0.19",
        sourceDir
      ]);
    }
  });

  logMessage("\n\n");
  displaySuccessMessage(Object.keys(packagesRequiringUpgrade));

  logMessage(
    "Here are some common upgrade steps that you will need to do manually:\n\n"
  );
  Object.keys(todos).forEach(function(groupName) {
    var messages = todos[groupName];
    if (messages.length <= 0) {
      return;
    }
    logMessage("- " + groupName + "\n");
    messages.forEach(function(message) {
      logMessage("  - [ ] " + message + "\n");
    });
  });
  logMessage("\n");
}

function init() {
  var got = require("got");
  var caw = require("caw");
  got(packageHost + "/search.json", { agent: caw() })
    .catch(function(err) {
      console.error(err);
      logError(
        "Unable to connect to " + packageHost + ".  Please try again later."
      );
      exit(1);
    })
    .then(function(response) {
      var knownPackages = JSON.parse(response.body);
      main(knownPackages);
    });
}

init();
